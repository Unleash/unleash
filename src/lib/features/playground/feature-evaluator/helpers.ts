import { IStrategyConfig } from '../../../types';
import { FeatureStrategiesEvaluationResult } from './client';
import { Context } from './context';

export type FallbackFunction = (name: string, context: Context) => boolean;

export function createFallbackFunction(
    name: string,
    context: Context,
    fallback?: FallbackFunction | boolean,
): () => FeatureStrategiesEvaluationResult {
    const createEvalResult = (enabled: boolean) => ({
        result: enabled,
        strategies: [],
    });

    if (typeof fallback === 'function') {
        return () => createEvalResult(fallback(name, context));
    }
    if (typeof fallback === 'boolean') {
        return () => createEvalResult(fallback);
    }
    return () => createEvalResult(false);
}

export function resolveContextValue(
    context: Context,
    field: string,
): string | undefined {
    if (context[field]) {
        return context[field] as string;
    }
    if (context.properties?.[field]) {
        return context.properties[field] as string;
    }
    return undefined;
}

export function safeName(str: string = ''): string {
    return str.replace(/\//g, '_');
}

export function getDefaultStrategy(featureName: string): IStrategyConfig {
    return {
        name: 'flexibleRollout',
        constraints: [],
        disabled: false,
        parameters: {
            rollout: '100',
            stickiness: 'default',
            groupId: featureName,
        },
    };
}

function resolveGroupId(
    defaultStrategy: IStrategyConfig,
    featureName: string,
): string {
    const groupId =
        defaultStrategy?.parameters?.groupId !== ''
            ? defaultStrategy.parameters?.groupId
            : featureName;

    return groupId || '';
}

export function getProjectDefaultStrategy(
    defaultStrategy: IStrategyConfig,
    featureName: string,
): IStrategyConfig {
    return {
        ...defaultStrategy,
        parameters: {
            ...defaultStrategy.parameters,
            groupId: resolveGroupId(defaultStrategy, featureName),
        },
    };
}
