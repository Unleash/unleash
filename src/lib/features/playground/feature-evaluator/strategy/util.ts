import * as murmurHash3 from 'murmurhash3js';

function normalizedValue(
    id: string,
    groupId: string,
    normalizer: number,
    seed = 0,
): number {
    const hash = murmurHash3.x86.hash32(`${groupId}:${id}`, seed);
    return (hash % normalizer) + 1;
}

const STRATEGY_SEED = 0;

export function normalizedStrategyValue(id: string, groupId: string): number {
    return normalizedValue(id, groupId, 100, STRATEGY_SEED);
}

const VARIANT_SEED = 86028157;

export function normalizedVariantValue(
    id: string,
    groupId: string,
    normalizer: number,
): number {
    return normalizedValue(id, groupId, normalizer, VARIANT_SEED);
}
