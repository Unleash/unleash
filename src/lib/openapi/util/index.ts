export * from './api-operation';
export * from './create-request-schema';
export * from './create-response-schema';
export * from './openapi-tags';
export * from './standard-responses';
