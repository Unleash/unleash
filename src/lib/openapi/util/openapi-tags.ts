// The "official" OpenAPI tags that we use. These tags are added to the OpenAPI
// spec as the root-level "tags" list. Consider creating a new entry here when
// creating a new endpoint.
const OPENAPI_TAGS = [
    {
        name: 'Addons',
        description:
            'Create, update, and delete [Unleash addons](https://docs.getunleash.io/addons).',
    },
    { name: 'Admin UI', description: 'Configure the Unleash Admin UI.' },
    {
        name: 'API tokens',
        description:
            'Create, update, and delete [Unleash API tokens](https://docs.getunleash.io/reference/api-tokens-and-client-keys).',
    },
    {
        name: 'Personal access tokens',
        description:
            'Create, update, and delete [Personal access tokens](https://docs.getunleash.io/reference/api-tokens-and-client-keys#personal-access-tokens).',
    },
    {
        name: 'Service Accounts',
        description:
            'Endpoints for managing [Service Accounts](https://docs.getunleash.io/reference/service-accounts), which enable programmatic access to the Unleash API.',
    },
    {
        name: 'Archive',
        description:
            'Revive or permanently delete [archived feature toggles](https://docs.getunleash.io/advanced/archived_toggles).',
    },
    { name: 'Auth', description: 'Manage logins, passwords, etc.' },
    {
        name: 'Client',
        description:
            'Endpoints for [Unleash server-side clients](https://docs.getunleash.io/reference/sdks).',
    },
    {
        name: 'Context',
        description:
            'Create, update, and delete [context fields](https://docs.getunleash.io/reference/unleash-context) that Unleash is aware of.',
    },
    {
        name: 'Environments',
        description:
            'Create, update, delete, enable or disable [environments](https://docs.getunleash.io/reference/environments) for this Unleash instance.',
    },
    { name: 'Events', description: 'Read events from this Unleash instance.' },
    {
        name: 'Features',
        description:
            'Create, update, and delete [features toggles](https://docs.getunleash.io/reference/feature-toggles).',
    },
    {
        name: 'Import/Export',
        description:
            '[Import and export](https://docs.getunleash.io/deploy/import_export) the state of your Unleash instance.',
    },
    {
        name: 'Instance Admin',
        description:
            'Instance admin endpoints used to manage the Unleash instance itself.',
    },
    {
        name: 'Metrics',
        description: 'Register, read, or delete metrics recorded by Unleash.',
    },
    {
        name: 'Operational',
        description:
            'Endpoints related to the operational status of this Unleash instance.',
    },
    {
        name: 'Playground',
        description:
            'Evaluate an Unleash context against your feature toggles.',
    },
    {
        name: 'Projects',
        description:
            'Create, update, and delete [Unleash projects](https://docs.getunleash.io/reference/projects).',
    },
    {
        name: 'Public signup tokens',
        description:
            'Create, update, and delete [Unleash Public Signup tokens](https://docs.getunleash.io/reference/public-signup-tokens).',
    },
    {
        name: 'Segments',
        description:
            'Create, update, delete, and manage [segments](https://docs.getunleash.io/reference/segments).',
    },
    {
        name: 'Strategies',
        description:
            'Create, update, delete, manage [custom strategies](https://docs.getunleash.io/reference/custom-activation-strategies).',
    },
    {
        name: 'Tags',
        description:
            'Create, update, and delete [tags and tag types](https://docs.getunleash.io/reference/tags).',
    },
    { name: 'Users', description: 'Manage users and passwords.' },
    {
        name: 'Unstable',
        description:
            'Experimental endpoints that may change or disappear at any time.',
    },
    { name: 'Edge', description: 'Endpoints related to Unleash on the Edge.' },
    {
        name: 'Frontend API',
        description:
            'API for connecting client-side (frontend) applications to Unleash.',
    },
    {
        name: 'Maintenance',
        description: 'Enable/disable the maintenance mode of Unleash.',
    },
    {
        name: 'Change Requests',
        description:
            'API for managing [change requests](https://docs.getunleash.io/reference/change-requests).',
    },
] as const;

// make the export mutable, so it can be used in a schema
export const openApiTags = [...OPENAPI_TAGS].sort((a, b) =>
    a.name.localeCompare(b.name),
);

export type OpenApiTag =
    // The official OpenAPI tags that we use.
    typeof OPENAPI_TAGS[number]['name'];
