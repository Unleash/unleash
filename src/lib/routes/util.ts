import joi from 'joi';
import { Response } from 'express';
import { Logger } from '../logger';
import BaseError from '../error/base-error';

export const customJoi = joi.extend((j) => ({
    type: 'isUrlFriendly',
    base: j.string(),
    messages: {
        'isUrlFriendly.base': '{{#label}} must be URL friendly',
    },
    validate(value, helpers) {
        // Base validation regardless of the rules applied
        if (encodeURIComponent(value) !== value) {
            // Generate an error, state and options need to be passed
            return { value, errors: helpers.error('isUrlFriendly.base') };
        }
        return undefined;
    },
}));

export const nameType = customJoi.isUrlFriendly().min(1).max(100).required();

export const handleErrors: (
    res: Response,
    logger: Logger,
    error: Error,
) => void = (res, logger, error) => {
    logger.warn(error.message);
    // @ts-expect-error
    // eslint-disable-next-line no-param-reassign
    error.isJoi = true;

    if (error instanceof BaseError) {
        return res.status(error.statusCode).json(error).end();
    }

    switch (error.name) {
        case 'ValidationError':
            return res.status(400).json(error).end();
        case 'BadDataError':
            return res.status(400).json(error).end();
        case 'OwaspValidationError':
            return res.status(400).json(error).end();
        case 'PasswordUndefinedError':
            return res.status(400).json(error).end();
        case 'MinimumOneEnvironmentError':
            return res.status(400).json(error).end();
        case 'InvalidTokenError':
            return res.status(401).json(error).end();
        case 'NoAccessError':
            return res.status(403).json(error).end();
        case 'UsedTokenError':
            return res.status(403).json(error).end();
        case 'InvalidOperationError':
            return res.status(403).json(error).end();
        case 'IncompatibleProjectError':
            return res.status(403).json(error).end();
        case 'OperationDeniedError':
            return res.status(403).json(error).end();
        case 'NotFoundError':
            return res.status(404).json(error).end();
        case 'NameExistsError':
            return res.status(409).json(error).end();
        case 'FeatureHasTagError':
            return res.status(409).json(error).end();
        case 'RoleInUseError':
            return res.status(400).json(error).end();
        case 'ProjectWithoutOwnerError':
            return res.status(409).json(error).end();
        default:
            logger.error('Server failed executing request', error);
            return res.status(500).end();
    }
};
