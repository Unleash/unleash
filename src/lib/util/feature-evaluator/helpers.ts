import { IStrategyConfig } from '../../types/model';
import { FeatureStrategiesEvaluationResult } from './client';
import { Context } from './context';

export type FallbackFunction = (name: string, context: Context) => boolean;

export function createFallbackFunction(
    name: string,
    context: Context,
    fallback?: FallbackFunction | boolean,
): () => FeatureStrategiesEvaluationResult {
    const createEvalResult = (enabled: boolean) => ({
        result: enabled,
        strategies: [],
    });

    if (typeof fallback === 'function') {
        return () => createEvalResult(fallback(name, context));
    }
    if (typeof fallback === 'boolean') {
        return () => createEvalResult(fallback);
    }
    return () => createEvalResult(false);
}

export function resolveContextValue(
    context: Context,
    field: string,
): string | undefined {
    if (context[field]) {
        return context[field] as string;
    }
    if (context.properties && context.properties[field]) {
        return context.properties[field] as string;
    }
    return undefined;
}

export function safeName(str: string = ''): string {
    return str.replace(/\//g, '_');
}

export function getDefaultStrategy(featureName: string): IStrategyConfig {
    return {
        name: 'flexibleRollout',
        constraints: [],
        parameters: {
            rollout: '100',
            stickiness: 'default',
            groupId: featureName,
        },
    };
}
