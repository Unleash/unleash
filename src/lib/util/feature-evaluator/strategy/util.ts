import * as murmurHash3 from 'murmurhash3js';

export default function normalizedValue(id: string, groupId: string, normalizer = 100): number {
  return (murmurHash3.x86.hash32(`${groupId}:${id}`) % normalizer) + 1;
}
