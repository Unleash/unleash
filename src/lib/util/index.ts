export * from './timer';
export * from './semver';
export * from './version';
export * from './is-email';
export * from './segments';
export * from './anonymise';
export * from './constants';
export * from './isDefined';
export * from './omit-keys';
export * from './random-id';
export * from './snakeCase';
export * from './map-values';
export * from './time-utils';
export * from './collect-ids';
export * from './parseEnvVar';
export * from './rewriteHTML';
export * from './extract-user';
export * from './flag-resolver';
export * from './metrics-helper';
export * from './validateOrigin';
export * from './anyEventEmitter';
export * from './format-base-uri';
export * from './load-index-html';
export * from './findPublicFolder';
export * from './generateImageUrl';
export * from './ensureStringValue';
export * from './graceful-shutdown';
export * from './validateJsonString';
export * from './arraysHaveSameItems';
export * from './constantTimeCompare';
export * from './collapseHourlyMetrics';
export * from './offline-unleash-client';
