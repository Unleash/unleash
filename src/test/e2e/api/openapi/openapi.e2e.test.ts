import { setupApp } from '../../helpers/test-helper';
import dbInit from '../../helpers/database-init';
import getLogger from '../../../fixtures/no-logger';
import SwaggerParser from '@apidevtools/swagger-parser';
import enforcer from 'openapi-enforcer';
import semver from 'semver';
import { openApiTags } from '../../../../lib/openapi/util/openapi-tags';

let app;
let db;

beforeAll(async () => {
    db = await dbInit('openapi', getLogger);
    app = await setupApp(db.stores);
});

afterAll(async () => {
    await app.destroy();
    await db.destroy();
});

test('should serve the OpenAPI UI', async () => {
    return app.request
        .get('/docs/openapi/')
        .expect('Content-Type', /html/)
        .expect(200)
        .expect((res) => expect(res.text).toMatchSnapshot());
});

test('should serve the OpenAPI spec', async () => {
    return app.request
        .get('/docs/openapi.json')
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((res) => {
            // Don't use the version field in snapshot tests. Having the version
            // listed in automated testing causes issues when trying to deploy
            // new versions of the API (due to mismatch between new tag versions etc).
            delete res.body.info.version;

            // This test will fail whenever there's a change to the API spec.
            // If the change is intended, update the snapshot with `jest -u`.
            expect(res.body).toMatchSnapshot();
        });
});

test('should serve the OpenAPI spec with a `version` property', async () => {
    return app.request
        .get('/docs/openapi.json')
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((res) => {
            const { version } = res.body.info;
            // ensure there's no whitespace or leading `v`
            expect(semver.clean(version)).toStrictEqual(version);

            // ensure the version listed is valid semver
            expect(semver.parse(version, { loose: false })).toBeTruthy();
        });
});

test('the generated OpenAPI spec is valid', async () => {
    const { body } = await app.request
        .get('/docs/openapi.json')
        .expect('Content-Type', /json/)
        .expect(200);

    // this throws if the swagger parser can't parse it correctly
    // also parses examples, but _does_ do some string coercion in examples
    try {
        await SwaggerParser.validate(body);
    } catch (err) {
        console.error(err);
        // there's an error here, so let's exit after showing it in the console.
        expect(true).toBe(false);
    }

    const [, enforcerError, enforcerWarning] = await enforcer(body, {
        fullResult: true,
        componentOptions: {
            exceptionSkipCodes: [
                // allow non-standard formats for strings (including 'uri')
                'WSCH001',

                // Schemas with an indeterminable type cannot serialize,
                // deserialize, or validate values. [WSCH005]
                //
                // This allows specifying the 'any' type for schemas (such as the
                // patchSchema)
                'WSCH005',
            ],
        },
    });

    if (enforcerWarning !== undefined) {
        console.warn(enforcerWarning);
    }
    if (enforcerError !== undefined) {
        console.error(enforcerError);
    }

    expect(enforcerWarning ?? enforcerError).toBe(undefined);
});

test('all root-level tags are "approved tags"', async () => {
    const { body: spec } = await app.request
        .get('/docs/openapi.json')
        .expect('Content-Type', /json/)
        .expect(200);

    const specTags = spec.tags;
    const approvedTags = openApiTags;

    // expect spec tags to be a subset of the approved tags
    expect(approvedTags).toEqual(expect.arrayContaining(specTags));
});
// All tags that are used for OpenAPI path operations must also be listed in the
// OpenAPI root-level tags list. For us, there's two immediate things that make
// this important:
//
// 1. Swagger UI groups operations by tags. To make sure that endpoints are
// listed where users would expect to find them, they should be given an
// appropriate tag.
//
// 2. The OpenAPI/docusaurus integration we use does not generate documentation
// for paths whose tags are not listed in the root-level tags list.
//
// If none of the official tags seem appropriate for an endpoint, consider
// creating a new tag.
test('all tags are listed in the root "tags" list', async () => {
    const { body: spec } = await app.request
        .get('/docs/openapi.json')
        .expect('Content-Type', /json/)
        .expect(200);

    const rootLevelTagNames = new Set(spec.tags.map((tag) => tag.name));

    // dictionary of all invalid tags found in the spec
    let invalidTags = {};
    for (const [path, data] of Object.entries(spec.paths)) {
        for (const [operation, opData] of Object.entries(data)) {
            // ensure that the list of tags for every operation is a subset of
            // the list of tags defined on the root level

            // check each tag for this operation
            for (const tag of opData.tags) {
                if (!rootLevelTagNames.has(tag)) {
                    // store other invalid tags that already exist on this
                    // operation
                    const preExistingTags =
                        (invalidTags[path] ?? {})[operation]?.invalidTags ?? [];

                    // add information about the invalid tag to the invalid tags
                    // dict.
                    invalidTags = {
                        ...invalidTags,
                        [path]: {
                            ...invalidTags[path],
                            [operation]: {
                                operationId: opData.operationId,
                                invalidTags: [...preExistingTags, tag],
                            },
                        },
                    };
                }
            }
        }
    }

    if (Object.keys(invalidTags).length) {
        // create a human-readable list of invalid tags per operation
        const msgs = Object.entries(invalidTags).flatMap(([path, data]) =>
            Object.entries(data).map(
                ([operation, opData]) =>
                    `${operation.toUpperCase()} ${path} (operation id: ${
                        opData.operationId
                    }) has the following invalid tags: ${opData.invalidTags
                        .map((tag) => `"${tag}"`)
                        .join(', ')}`,
            ),
        );

        // format message
        const errorMessage = `The OpenAPI spec contains path-level tags that are not listed in the root-level tags object. The relevant paths, operation ids, and tags are as follows:\n\n${msgs.join(
            '\n\n',
        )}\n\nFor reference, the root-level tags are: ${spec.tags
            .map((tag) => `"${tag.name}"`)
            .join(', ')}`;

        console.error(errorMessage);
    }
    expect(invalidTags).toStrictEqual({});
});
