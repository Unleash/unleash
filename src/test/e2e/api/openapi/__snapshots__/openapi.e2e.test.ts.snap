// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should serve the OpenAPI UI 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <title>Swagger UI</title>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <style>
      html {
          box-sizing: border-box;
          overflow: -moz-scrollbars-vertical;
          overflow-y: scroll;
      }
      *,
      *:before,
      *:after {
          box-sizing: inherit;
      }
      body {
        margin: 0;
        padding: 0;
        background: #fafafa;
      }
    </style>
    
      <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"./swagger-ui.css\\" >
    
  </head>
  <body>
    
      <div id=\\"swagger-ui\\"></div>
      <script src=\\"./swagger-ui-bundle.js\\"></script>
      <script src=\\"./swagger-ui-standalone-preset.js\\"></script>
      <script>
        window.onload = function () {
          window.ui = SwaggerUIBundle({
            url: \\"/docs/openapi.json\\",
            dom_id: '#swagger-ui'
          })
        }
      </script>
    
  </body>
</html>
  "
`;

exports[`should serve the OpenAPI spec 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "constraintSchema": Object {
        "additionalProperties": false,
        "properties": Object {
          "contextName": Object {
            "type": "string",
          },
          "operator": Object {
            "type": "string",
          },
          "values": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "contextName",
          "operator",
        ],
        "type": "object",
      },
      "createFeatureSchema": Object {
        "properties": Object {
          "description": Object {
            "type": "string",
          },
          "impressionData": Object {
            "type": "boolean",
          },
          "name": Object {
            "type": "string",
          },
          "type": Object {
            "type": "string",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
      "featureSchema": Object {
        "additionalProperties": false,
        "properties": Object {
          "createdAt": Object {
            "format": "date",
            "nullable": true,
            "type": "string",
          },
          "description": Object {
            "type": "string",
          },
          "enabled": Object {
            "type": "boolean",
          },
          "impressionData": Object {
            "type": "boolean",
          },
          "lastSeenAt": Object {
            "format": "date",
            "nullable": true,
            "type": "string",
          },
          "name": Object {
            "type": "string",
          },
          "project": Object {
            "type": "string",
          },
          "stale": Object {
            "type": "boolean",
          },
          "strategies": Object {
            "items": Object {
              "$ref": "#/components/schemas/strategySchema",
            },
            "type": "array",
          },
          "type": Object {
            "type": "string",
          },
          "variants": Object {
            "items": Object {
              "$ref": "#/components/schemas/variantSchema",
            },
            "type": "array",
          },
        },
        "required": Array [
          "name",
          "project",
        ],
        "type": "object",
      },
      "featuresSchema": Object {
        "additionalProperties": false,
        "properties": Object {
          "features": Object {
            "items": Object {
              "$ref": "#/components/schemas/featureSchema",
            },
            "type": "array",
          },
          "version": Object {
            "type": "integer",
          },
        },
        "required": Array [
          "version",
          "features",
        ],
        "type": "object",
      },
      "overrideSchema": Object {
        "additionalProperties": false,
        "properties": Object {
          "contextName": Object {
            "type": "string",
          },
          "values": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
        },
        "required": Array [
          "contextName",
          "values",
        ],
        "type": "object",
      },
      "strategySchema": Object {
        "additionalProperties": false,
        "properties": Object {
          "constraints": Object {
            "items": Object {
              "$ref": "#/components/schemas/constraintSchema",
            },
            "type": "array",
          },
          "id": Object {
            "type": "string",
          },
          "name": Object {
            "type": "string",
          },
          "parameters": Object {
            "type": "object",
          },
        },
        "required": Array [
          "id",
          "name",
          "constraints",
          "parameters",
        ],
        "type": "object",
      },
      "variantSchema": Object {
        "additionalProperties": false,
        "properties": Object {
          "name": Object {
            "type": "string",
          },
          "overrides": Object {
            "items": Object {
              "$ref": "#/components/schemas/overrideSchema",
            },
            "type": "array",
          },
          "payload": Object {
            "type": "object",
          },
          "stickiness": Object {
            "type": "string",
          },
          "weight": Object {
            "type": "number",
          },
          "weightType": Object {
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "weight",
          "weightType",
          "stickiness",
          "overrides",
        ],
        "type": "object",
      },
    },
    "securitySchemes": Object {
      "apiKey": Object {
        "in": "header",
        "name": "Authorization",
        "type": "apiKey",
      },
    },
  },
  "info": Object {
    "title": "Unleash API",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/api/admin/archive/features": Object {
      "get": Object {
        "deprecated": true,
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/featuresSchema",
                },
              },
            },
            "description": "featuresResponse",
          },
        },
        "tags": Array [
          "admin",
        ],
      },
    },
    "/api/admin/features/": Object {
      "get": Object {
        "deprecated": true,
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/featuresSchema",
                },
              },
            },
            "description": "featuresResponse",
          },
        },
        "tags": Array [
          "admin",
        ],
      },
    },
    "/api/admin/projects/{projectId}/features": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "projectId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/featuresSchema",
                },
              },
            },
            "description": "featuresResponse",
          },
        },
        "tags": Array [
          "admin",
        ],
      },
      "post": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "projectId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/createFeatureSchema",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/featureSchema",
                },
              },
            },
            "description": "featureResponse",
          },
        },
        "tags": Array [
          "admin",
        ],
      },
    },
    "/api/admin/projects/{projectId}/features/{featureName}": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "projectId",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
          Object {
            "in": "path",
            "name": "featureName",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/featureSchema",
                },
              },
            },
            "description": "featureResponse",
          },
        },
        "tags": Array [
          "admin",
        ],
      },
    },
  },
  "security": Array [
    Object {
      "apiKey": Array [],
    },
  ],
  "servers": Array [
    Object {
      "url": "http://localhost:4242",
    },
  ],
}
`;
